ENH: add `optimization` to `QMCEngine`

tupui opened the pull request:
Generalize the parameter `optimization` for all `QMCEngine`.

* Factory function to get the optimization function,
* Left the possibility to easily have options as a parameter as well,
* Add doc,
* Add a smock test going through all options for all engines (optimization methods are already tested elsewhere),
* Enable `lloyd`.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -705,23 +720,53 @@ def __init__(
+            "rng": self.rng,
+
+            # lloyd
+            "tol":  1e-5,
+            "maxiter":  10,
```
Fix lint issues
```suggestion
            "tol": 1e-5,
            "maxiter": 10,
```

tupui commented:
@mdhaber will you have some time to look at this before the branching?

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -611,6 +612,20 @@ class QMCEngine(ABC):
+          design shows better space filling robustness toward 2D and 3D
+          subprojections compared to using other discrepancy measures.
+        * ``lloyd``: Perturb samples using Lloyd-Max algorithm. The process
+          converges to equally spaced samples.
```
Currently, we have "After optimization, the result would not be guaranteed to be of strength 2.", but now it would be that after optimization with method = 'lloyd', the result is no longer guaranteed to be a LHS at all, right?

mdhaber replied:
Then we should mention it.

tupui replied:
Done

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1109,66 +1172,31 @@ class LatinHypercube(QMCEngine):
                        f" of {set(lhs_method_strength)!r}")
             raise ValueError(message) from exc
 
-        optimization_method: Dict[Literal["random-cd"], Callable] = {
```
This is all moved to QMCEngine or something, I imagine? What changes are needed?

tupui replied:
This part was moved to `_select_optimizer` which indeed is used in `QMCEngine`. This way I have a clear separation of responsibilities and I don't have a gigantic init. This way it's easier to read I think.

mdhaber commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -581,6 +580,11 @@ def test_distribution(self, scramble):
         )
 
+    @pytest.mark.parametrize("optimization", [None, "random-CD", "lloyd"])
+    def test_select_optimizer(self, optimization):
```
Should probably add test for invalid value.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -2025,7 +2108,7 @@ def lloyd_centroidal_voronoi_tessellation(
     l1_old = _l1_norm(sample=sample)
     for i in range(maxiter):
-        sample = _lloyd_centroidal_voronoi_tessellation(
+        sample = _centroidal_voronoi_tessellation(
```
Before we had `lloyd_centroidal_voronoi_tessellation` and the internal private method `_ lloyd_centroidal_voronoi_tessellation`. In the previous PR, you suggested we should make this private in the name as well. Hence I renamed `lloyd_centroidal_voronoi_tessellation` to `_lloyd_centroidal_voronoi_tessellation` and I had to find a new name for the internal private method. I chose `_centroidal_voronoi_tessellation`

mdhaber replied:
Haha ok. 
Would something involving `_iteration` (e.g. `_lloyd_iteration` or `_lloyd_centroidal_voronoi_tessellation_iteration` or `_centroidal_voronoi_tessellation_iteration`) help describe the distinction between the two?

tupui replied:
If you prefer sure. The `lloyd` part is conveying this normally but I don't care too much 😅

tupui replied:
Ok I will update to `_lloyd_centroidal_voronoi_tessellation` and `_lloyd_iteration`.

tupui replied:
Done

tupui commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -581,6 +580,11 @@ def test_distribution(self, scramble):
         )
 
+    @pytest.mark.parametrize("optimization", [None, "random-CD", "lloyd"])
+    def test_select_optimizer(self, optimization):
```
We have one in the LHS tests. I can move it to here to make it clearer.

mdhaber replied:
I think so. We're moving `optimization` to the level of `QMCEngine`, so I think the test should move to the level of `QMCEngine`, too.

tupui replied:
Ok sure will do.

tupui commented on scipy/stats/tests/test_qmc.py:
```diff
@@ -581,6 +580,11 @@ def test_distribution(self, scramble):
+    @pytest.mark.parametrize("optimization", [None, "random-CD", "lloyd"])
+    def test_select_optimizer(self, optimization):
+        engine = self.engine(d=2, optimization=optimization, scramble=False)
+        engine.random(32)
```
Yes in the LHS tests. And lloyd has it's own tests.

Here and with the error message test, I tried not to change all tests so we are confident with our changes. OC I can make the changes.

tupui replied:
Yes OC adding more test would make the test suite stronger. I just thought it was not that necessary. i.e. it does not really matter which method you use it works the same way: some permutation of the samples and only accept a new one if there is an improvement vs a metric.

If you prefer I can move the test which is in LHS and check for all engines and optimization methods.

mdhaber replied:
Yup, that sounds great.

tupui replied:
Done

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -705,23 +720,53 @@ def __init__(
         n : int, optional
             Number of samples to generate in the parameter space.
             Default is 1.
+        workers : int, optional
```
This param was (and still is) only available for `Halton` before and now because the `random` is shared, it has to be documented there. From a design perspective it's also better to have the same interface no matter the child class (IDE was complaining about this actually).

tupui reviewed (commented):
Thanks for the review @mdhaber. I will add your changes and update the PR since last time I did (and CI was run), it was before I added `PoissonDisk`.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -1304,6 +1284,22 @@ class Sobol(QMCEngine):
         ``np.float64`` to prevent points from repeating themselves.
         Default is None, which for backward compatibility, corresponds to 30.
+
+        .. versionadded:: 1.9.0
```
No this one relates to `bits` which will be in 1.9

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -611,6 +612,20 @@ class QMCEngine(ABC):
+          centered discrepancy. The best design based on the centered
+          discrepancy is constantly updated. Centered discrepancy-based
+          design shows better space filling robustness toward 2D and 3D
+          subprojections compared to using other discrepancy measures.
```
Yes it can help. I measured some convergence improvement on Sobol' for instance and it would also help Halton the same way. In the end it's another way to scramble. For Sobol' specifically, Sobol' has the sub projection properties of LHS. It is a LHS. So it does make as much sense as with LHS. I am just not sure it preserves all properties, I could try to check this point. In any case I think it has some value.

mdhaber replied:
OK. It would be great if you'd show a bit of this in the PR description (e.g. a short snippet we can run to show the benefits of these changes for the various `QMCEngine`s).

tupui replied:
I actually did this for Sobol'. Permutations would improve a bit the error (starting from unscrambled). It's not doing better than the fancy scrambling method, but it's an improvement over the stock version.

<img width="930" alt="Screenshot 2022-06-16 at 16 58 25" src="https://user-images.githubusercontent.com/23188539/174099386-ba1ac9a4-9423-4fb1-ba3f-e04ad346a1d4.png">


Diving deeper into the topic, we would need to use other criteria than the discrepancy to optimize. Art et al. are always debating on what to use. Disrepancy is good but some say it's incomplete, etc. There are other fancy methods but even more spooky to use/tune.

In the end, I think the value here is that we have an easy way to perturb our engines' samples without breaking (and even improving on some metric) too much things.

mdhaber replied:
I think it's fine to just improve the discrepancy since, at least for the random method, that's exactly what it's supposed to do.
Can I play with that code? It looks like the red line is for unscrambled + optimization, but what about scrambled + optimization?

tupui replied:
Sure here is the code I am using when doing such things. I actually did not keep this specific experiment though. But easy to add as you will see.

> It looks like the red line is for unscrambled + optimization, but what about scrambled + optimization?

Yes, for the scrambled + optimization it increases the error and you end up almost with the same red line as now. The scrambling in Sobol' is just better and very hard to beat. 

<details>
 <summary>Convergence on Sobol'</summary>

```python
import os
from collections import namedtuple

import numpy as np
from scipy.stats import qmc

import matplotlib.pyplot as plt
from matplotlib.container import ErrorbarContainer
from matplotlib.legend_handler import HandlerErrorbar

# Parameters
dim = 5
n_conv = 99

MIN_POWER = 4
POWER_MAX = 21
ns_gen = 2 ** np.arange(MIN_POWER, POWER_MAX + 1)  # 13

path = f'sobol_convergence'
os.makedirs(path, exist_ok=True)


# Functions definitions
_exp1 = 1 - np.exp(1)


def art_0(sample):
    # dim 2, true value 0
    return np.prod(sample - 0.5, axis=1)


def art_1d(sample):
    # dim 1, true value 1
    return sample * np.exp(sample)


def art_1(sample):
    # dim 5, true value 0
    return np.sum(np.exp(sample) + _exp1, axis=1)


def art_2(sample):
    # dim 5, true value 5/3 + 5*(5 - 1)/4
    return np.sum(sample, axis=1) ** 2


def art_3(sample):
    # dim 3, true value 0
    return np.prod(np.exp(sample) + _exp1, axis=1)


def type_a(sample, dim=30):
    # true value 1
    a = np.arange(1, dim + 1)
    f = 1.
    for i in range(dim):
        f *= (abs(4. * sample[:, i] - 2) + a[i]) / (1. + a[i])
    return f


def type_b(sample, dim=30):
    # true value 1
    f = 1.
    for d in range(1, dim + 1):
        f *= (d - sample[:, d - 1]) / (d - 0.5)
    return f


def type_b_f2(sample, dim=30):
    # true value 1
    f = 1.
    inv = 1. / dim
    for d in range(dim):
        f *= (1. + 1. / dim) * sample[:, d] ** inv
    return f


def type_c(sample, dim=10):
    # true value 1
    f = 2 ** dim * np.prod(sample, axis=1)
    return f


def additive(sample):
    # true value 1/2*dim
    return np.sum(sample, axis=1)


def pi(sample):
    # true value 1/2*dim
    return np.sum(sample, axis=1)


functions = namedtuple('functions', ['name', 'func', 'dim', 'ref'])

benchmark = [
    functions('Art 2D', art_0, 2, 0),
    functions('Art 1D', art_1d, 1, 1),
    functions('Art 1', art_1, 5, 0),
    functions('Art 2', art_2, 5, 5 / 3 + 5 * (5 - 1) / 4),
    functions('Art 3', art_3, 3, 0),
    functions('Type A', type_a, 30, 1),
    functions('Type B f3', type_b, 30, 1),
    functions('Type B f2', type_b_f2, 30, 1),
    functions('Type C', type_c, 10, 1),
    functions('Additive', additive, 30, 1/2*30)
]


def conv_method(samples, func, n_samples, n_conv, ref):
    evals = [np.sum(func(sample)) / n_samples for sample in samples]
    squared_errors = (ref - np.array(evals)) ** 2
    rmse = (np.sum(squared_errors) / n_conv) ** 0.5

    return rmse, 0.  # can add another metric


# Analysis

def sampler_mc(ns, dim):
    sampler_ = lambda x: np.random.random((x, dim))
    samples_ = [sampler_(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=False)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_scrambled_13bits(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=True, bits=13)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_scrambled_16bits(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=True, bits=16)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_scrambled_32bits(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=True, bits=32)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


def sampler_sobol_scrambled_64bits(ns, dim):
    def _sampler_sobol(ns):
        engine = qmc.Sobol(d=dim, scramble=True, bits=64)
        return engine.random(ns)

    samples_ = [_sampler_sobol(ns) for _ in range(n_conv)]
    return np.array(samples_)


samplers = {
    "MC": [sampler_mc, []],
    "Sobol' unscrambled": [sampler_sobol, []],
    # "Sobol' scrambled 13 bits": [sampler_sobol_scrambled_13bits, []],
    # "Sobol' scrambled 16 bits": [sampler_sobol_scrambled_16bits, []],
    "Sobol' scrambled 32 bits": [sampler_sobol_scrambled_32bits, []],
    "Sobol' scrambled 64 bits": [sampler_sobol_scrambled_64bits, []],
}


for ns in ns_gen:
    print(f'-> ns={ns}')

    for sampler_ in samplers:
        sample_ = samplers[sampler_][0](ns, dim)

        conv_res = []
        for case in benchmark:
            if case.dim != dim:
                continue

            conv_res_ = conv_method(sample_, case.func, ns, n_conv, case.ref)
            conv_res.append(conv_res_)

        samplers[sampler_][1].append(conv_res)


for sampler_ in samplers:
    samplers[sampler_][1] = np.array(samplers[sampler_][1])


# Plot
i = -1
for case in benchmark:
    if case.dim != dim:
        continue
    i += 1

    func = case.name
    fig, ax = plt.subplots()

    # ratio_1 = sample_sobol_0_rmse[:, i, 0][0] / ns_gen[0] ** (-2/2)
    # ratio_1 = sample_sobol_scramble_rmse[:, i, 0][0] / ns_gen[0] ** (-2 / 2)
    # ratio_2 = sample_sobol_scramble_0_rmse[:, i, 0][0] / (np.log2(ns_gen[0]) * ns_gen[0] ** (-3 / 2))
    # ratio_3 = sample_sobol_scramble_0_rmse[:, i, 0][0] / (ns_gen[0] ** (-3 / 2))

    # ax.plot(ns_gen, ns_gen ** (-1 / 2), ls='-', c='k')
    # ax.plot(ns_gen, ns_gen ** (-2/2) * ratio_1, ls='-', c='k')
    # ax.plot(ns_gen, np.log2(ns_gen) * ns_gen ** (-3 / 2) * ratio_2, ls='-.')
    # ax.plot(ns_gen, ns_gen ** (-3 / 2) * ratio_3, ls='-', c='k')

    for sampler_ in samplers:
        ax.plot(
            ns_gen, np.log2(samplers[sampler_][1][:, i, 0]),
            label=sampler_
        )

    ax.set_xlabel(r'$N_s$')

    ax.set_xscale('log')
    # ax.set_yscale('log')

    ax.set_xticks(ns_gen)
    ax.set_xticklabels([fr'$2^{{{ns}}}$' for ns in np.arange(MIN_POWER, POWER_MAX + 1)])
    ax.set_ylabel(r'$\log (\epsilon (\mu))$')
    fig.legend(labelspacing=0.7, bbox_to_anchor=(0.5, 0.45),
               handler_map={ErrorbarContainer: HandlerErrorbar(xerr_size=0.7)})
    fig.tight_layout()
    # plt.show()
    fig.savefig(os.path.join(path, f'sobol_conv_integration_mean_{func}.pdf'),
                transparent=True, bbox_inches='tight')

```

</details>

mdhaber replied:
K. At first glance, it looks like the "error" referred to above is not centered discrepancy, as I had assumed. It is one of the "functions" defined toward the top (e.g. `additive`)?
 
I understand that there is debate about what metric to use to assess the QMC quality. Whatever metric is used here and in the tests should show that `optimize` has an advanctage over `scrambled`. (Otherwise, the question would arise: "what is the purpose of adding `optimize` if `scrambled` - the default - is better?") 

So maybe since `random-cd` is described with "Centered discrepancy-based design shows better space filling robustness toward 2D and 3D subprojections", choose a metric that shows that `random-cd` gives better space-filling robustness. Since `lloyd` "converges to equally spaced samples", choose a metric that values equally spaced samples.

Does that make sense, or am I still missing the point? Is there some other reason to add `optimize` even if the default `scrambled=True` is superior according to the metric shown here?

mdhaber replied:
This is what I had in mind:
```python3
import numpy as np
from scipy.stats import qmc
import matplotlib.pyplot as plt

QMCEngine = qmc.LatinHypercube
metric = qmc.discrepancy
metric_options = {'method': 'CD'}
ns = 2**np.arange(4, 11).astype(int)  # sample sizes
optimizers = (None, 'random-cd', 'lloyd')

results = {f"optimizer={optimizer}":[] for optimizer in optimizers}

for n in ns:
    for optimizer in optimizers:
        rng = np.random.default_rng(23982491)
        qrng = QMCEngine(d=2, seed=rng, optimization=optimizer)
        sample = qrng.random(n=n)
        val = metric(sample=sample, **metric_options)
        results[f"optimizer={optimizer}"].append(val)

for name, data in results.items():
    plt.loglog(ns, data)
plt.legend(results.keys())
plt.xlabel("Size of Sample")
plt.ylabel(f"{metric.__name__} with {metric_options}")
plt.title(f"Effectiveness of Optimization for {QMCEngine.__name__}")
```
This is expected.
![image](https://user-images.githubusercontent.com/6570539/174392568-f58e93fa-8a52-4bda-8313-6765e55f023d.png)

But I'm not seeing much benefit to the CD of `Halton` or `Sobol`:
![image](https://user-images.githubusercontent.com/6570539/174392707-e17b42db-ef77-47dd-940f-38383a1b2fb6.png)

For what metrics should I expect to see substantial benefit?

tupui replied:
Cool thanks 👍 For Lloyd it is somehow expected that CD drops as it is optimizing the L2 instead. L2 and CD are linked but not equivalent and lead to very different designs in practice. I would not try to compare Lloyd with CD directly because of that.

In the end it really depends on the problem you are trying to solve. My plots was about the error you make on the mean. It's often what you look for, but you can be constrained.

E.g. if you need to respect a minimal L2, then Lloyd would be a good option (I had this once, my parameters represented some geometrical parameters of an object designed with additive manufacturing. I wanted to make sure that another sample was making enough geometrical modification of the object so that if you were to manufacture all samples, you would actually have a deviation in the design. I was simulating all that in my case, making CFD simulation of all the samples.).

mdhaber replied:
Could you write out the "minimal L2" function so we use it as the metric in the plot and see the advantage of Lloyd?

tupui replied:
We have it at the end of the test file. This is how we check that Lloyd is working.

```python
def l2_norm(sample):
    return distance.pdist(sample).min()
```

mdhaber replied:
OK, so by "minimal L2", you meant that we want it to "maximize the minimum of the pairwise L2" distances, right? (Sorry I don't remember these things. As you know, we work on a lot of things.)

![image](https://user-images.githubusercontent.com/6570539/175828382-ee381c5e-5c1c-4e8d-8533-71bfbf9236e9.png)

So `lloyd` does not appear to be doing its job for some sample sizes, right?

> E.g. if you need to respect a minimal L2, then Lloyd would be a good option 

It sounds like the idea is to get somewhat random but well-spaced points? Are there any other examples that suggest that others would want to use QMC (as opposed to PRNG samples) to achieve this?

tupui replied:
> OK, so by "minimal L2", you meant that we want it to "maximize the minimum of the pairwise L2" distances, right? (Sorry I don't remember these things. As you know, we work on a lot of things!)

No worries. Yes this is what I mean, maximise the inter-sample distance (L2 or L1). As you saw with the second trial, it can happen that the metric is worsen for a few iteration before improving again (as opposed to `random-cd`, Lloyd does not accept nor reject a sample based on a metric. It only monitors the L1 as a stopping criterion). Aside from pathological cases, running a bit longer the algorithm should improve the metric.

tupui replied:
> It sounds like the idea is to get somewhat random but well-spaced points?

Exactly.

> Are there any other examples that suggest that others would want to use QMC (as opposed to PRNG samples) to achieve this?

Not sure I correctly get the question. But using QMC vs PRNG samples as a starting point to get well-spaced points would be faster since QMC already space well points in the space. In high dimensions, it would be more difficult/timely to break clusters than using a sample which has no clustering by construction. Samples need to move "less" to get to a satisfactory distance from each others. Also, since QMC has better convergence properties than PRNG, it's natural to look for them in any cases.

mdhaber replied:
> Aside from pathological cases, running a bit longer the algorithm should improve the metric.

The plot shows several cases in which the metric was apparently worsened, not improved, by lloyd. These plots do not show the metric at each iteration; they show the metric on the output of `random` (with different `optimizer` options) as a function of the sample size.

mdhaber replied:
I'll defer to your judgment here.

tupui commented:
I addressed your comments. While timing the tests, I also noticed that `TestPoisson::test_mindist` was the slowest test. I reduced a bit the test matrix and most importantly fixed the dimension which was not being tested.

mdhaber commented:
Ok, I'll take a final pass. I guess I'll leave the points about the effectiveness of Lloyd to your judgement. I'll merge if the changes look good.

mdhaber commented:
@tupui Looks like there are failures after `TestPoisson::test_mindist` was changed.

Otherwise, looks like most comments were addressed. We can leave the descriptions as they are, but I had some thoughts.

> random permutations of coordinates to lower the centered discrepancy. The best design based on the centered discrepancy is constantly updated. Centered discrepancy-based design shows better space filling robustness toward 2D and 3D subprojections compared to using other discrepancy measures.

The word "designs" may be less appropriate for engines other than LHS. Are there special properties of some of the engines that might be destroyed by these changes?

>  Perturb samples using a modified Lloyd-Max algorithm. The process converges to equally spaced samples.

Hopefully! Consider qualifying the statement, since it's not guaranteed.

tupui commented:
Ark, fixing the test makes it too large in memory. Before we were not testing for dim=4, only 2 and it was ok. I will make a fix.

> Hopefully! Consider qualifying the statement, since it's not guaranteed.

Will soften this at the same time.

> The word "designs" may be less appropriate for engines other than LHS.

I can change it for "sample" if you prefer. Design, as design of experiments (DoI), is very much used in the QMC community and does not really have additional meaning than "sample".

> Are there special properties of some of the engines that might be destroyed by these changes?

Sobol' is a LHS with a better arrangement. It would conserve this (permutation is done by selecting one column and permuting 2 rows. Points in a LHS/Sobol' are like a sudoku on a grid, if you do such permutation, you still have your sudoku). Halton does not guarantee anything, so nothing is broken.

mdhaber commented:
> I can change it for "sample" if you prefer.

I think `sample` is more general. 

> I should probably have a general statement saying that optimization comes at the cost of not strictly conserving all initial properties of the designs and are meant to help with specific problems.

That sounds good.

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -612,6 +613,21 @@ class QMCEngine(ABC):
+    optimization : {None, "random-cd", "lloyd"}, optional
+        Whether to use an optimization scheme to improve the quality after
+        sampling. Note that this is a post-processing step that does not
+        guarrantee that all properties of the sample would be conserved.
```
```suggestion
        guarantee that all properties of the sample will be conserved.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -612,6 +613,21 @@ class QMCEngine(ABC):
+        * ``random-cd``: random permutations of coordinates to lower the
+          centered discrepancy. The best sample based on the centered
+          discrepancy is constantly updated. Centered discrepancy-based
+          sample shows better space filling robustness toward 2D and 3D
```
```suggestion
          sampling shows better space-filling robustness toward 2D and 3D
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -853,6 +899,21 @@ class Halton(QMCEngine):
+    optimization : {None, "random-cd", "lloyd"}, optional
+        Whether to use an optimization scheme to improve the quality after
+        sampling. Note that this is a post-processing step that does not
+        guarrantee that all properties of the sample would be conserved.
```
```suggestion
        guarantee that all properties of the sample will be conserved.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -853,6 +899,21 @@ class Halton(QMCEngine):
+        * ``random-cd``: random permutations of coordinates to lower the
+          centered discrepancy. The best sample based on the centered
+          discrepancy is constantly updated. Centered discrepancy-based
+          sample shows better space filling robustness toward 2D and 3D
```
```suggestion
          sampling shows better space-filling robustness toward 2D and 3D
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -975,17 +1036,23 @@ class LatinHypercube(QMCEngine):
+    optimization : {None, "random-cd", "lloyd"}, optional
+        Whether to use an optimization scheme to improve the quality after
+        sampling. Note that this is a post-processing step that does not
+        guarrantee that all properties of the sample would be conserved.
```
```suggestion
        guarantee that all properties of the sample will be conserved.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -975,17 +1036,23 @@ class LatinHypercube(QMCEngine):
           discrepancy is constantly updated. Centered discrepancy-based
-          design shows better space filling robustness toward 2D and 3D
-          subprojections compared to using other discrepancy measures [6]_.
+          sample shows better space filling robustness toward 2D and 3D
```
```suggestion
          sampling shows better space-filling robustness toward 2D and 3D
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1305,6 +1289,23 @@ class Sobol(QMCEngine):
+    optimization : {None, "random-cd", "lloyd"}, optional
+        Whether to use an optimization scheme to improve the quality after
+        sampling. Note that this is a post-processing step that does not
+        guarrantee that all properties of the sample would be conserved.
```
```suggestion
        guarantee that all properties of the sample will be conserved.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1305,6 +1289,23 @@ class Sobol(QMCEngine):
+        * ``random-cd``: random permutations of coordinates to lower the
+          centered discrepancy. The best sample based on the centered
+          discrepancy is constantly updated. Centered discrepancy-based
+          sample shows better space filling robustness toward 2D and 3D
```
```suggestion
          sampling shows better space-filling robustness toward 2D and 3D
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1615,6 +1618,21 @@ class PoissonDisk(QMCEngine):
+    optimization : {None, "random-cd", "lloyd"}, optional
+        Whether to use an optimization scheme to improve the quality after
+        sampling. Note that this is a post-processing step that does not
+        guarrantee that all properties of the sample would be conserved.
```
```suggestion
        guarantee that all properties of the sample will be conserved.
```

mdhaber commented on scipy/stats/_qmc.py:
```diff
@@ -1615,6 +1618,21 @@ class PoissonDisk(QMCEngine):
+        * ``random-cd``: random permutations of coordinates to lower the
+          centered discrepancy. The best sample based on the centered
+          discrepancy is constantly updated. Centered discrepancy-based
+          sample shows better space filling robustness toward 2D and 3D
```
```suggestion
          sampling shows better space-filling robustness toward 2D and 3D
```

mdhaber reviewed (approved):
The `random-cd` option tends to lower the centered discrepancy a bit for all `QMCEngines`, and the `lloyd` method usually increases the minimum distance between points. I believe these are properties of samples that users might wish to improve, so it can be useful to have these options. I'll merge after the minor doc fixes pass the lint check.

mdhaber commented:
Failure seems unrelated.

tupui commented:
Thanks for all your help moving this forward @mdhaber 😃
